source/common/tracing/zipkin/zipkin_core_types.h

```diff
@@ -0,0 +1,586 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "common/tracing/zipkin/tracer_interface.h"
+#include "common/tracing/zipkin/util.h"
+
+namespace Zipkin {
+
+/**
+ * Base class to be inherited by all classes that represent Zipkin-related concepts, namely:
+ * endpoint, annotation, binary annotation, and span.
+ */
+class ZipkinBase {
+public:
+  /**
+   * Destructor.
+   */
+  virtual ~ZipkinBase() {}
+
+  /**
+   * All classes defining Zipkin abstractions need to implement this method to convert
+   * the corresponding abstraction to a Zipkin-compliant JSON.
+   */
+  virtual const std::string& toJson() PURE;
+
+protected:
+  std::string json_string_;
+};
+
+/**
+ * Represents a Zipkin endpoint. This class is based on Zipkin's Thrift definition of an endpoint.
+ * Endpoints can be added to Zipkin annotations.
+ */
+class Endpoint : public ZipkinBase {
```

Comment:
We already have a full set of interfaces for network addresses. Rework this to use Network::Address as a member. Then ipv4 and ipv6, etc. will be handled for you.